network/src/peers/peer/cipher.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::convert::TryInto;
+
+use snow::TransportState;
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+
+use crate::NetworkError;
+
+pub struct Cipher {
+    state: TransportState,
+    buffer: Box<[u8]>,
+    noise_buffer: Box<[u8]>,
+}
+
+impl Cipher {
+    pub fn new(state: TransportState, buffer: Box<[u8]>, noise_buffer: Box<[u8]>) -> Self {
+        assert_eq!(buffer.len(), crate::MAX_MESSAGE_SIZE);
```

Comment:
the read buffer can be greater than the protocol's max message size for performance reasons

Reply:
how does it help with performance? and its always allocated to be that size currently

Reply:
it might have a positive impact on performance in the future (since we could then read bytes from the stream faster, queuing them in memory), but at this point in time the important thing is that the `buffer` is at least as big as `crate::MAX_MESSAGE_SIZE`, i.e. `assert!(buffer.len() >= crate::MAX_MESSAGE_SIZE);` would be a better pick here

Reply:
that won't have a performance improvement here. we could use a buffered read stream, yeah, but right now its read exact length packets, and because we don't do (a lot) of small reads, buffered read stream is probably negligible. this buffer is just for incoming packets, not for a general read buffer.